Various improvements

alexander-veit opened the pull request:
This PR introduces various bug fixes and new features.

- Added small example files to check non-standard reads (deletions, soft/hard clipping)
- Build out `es` modules when "npm run build" is executed
- Add option `workerScriptLocation`. This is needed when the worker script is not in the same folder as the pileup script
- The client side data fetcher now accepts `bamUrl` and `baiUrl`. `url` is still supported
- Added options `plusStrandColor` and `minusStrandColor`. When set, the reads are colored according to their strand.
- Added the option `showReadCounts` and `readCountHeight,` which displays read count information (like a bar track) on top of the track
- Improved mouseovers for reads and read counts
- Bumped version of GMOD/bam-js, because the current version did not allow to retrieve the read sequence
- Variants are now correctly colored based on the read sequence.
- Added visualization of reads that contain hard clipping
- Fixed a bug where MD strings would not be correctly parsed, when a deletion is present, e.g., `85^A16`
- Added stripes to represent deletions. It's not extremely pretty, but it's easier to distinguish. Please let me know if you don't want that. There is an example in the `index.html`, that you can have a look at.
- When there is soft clipping at the beginning of a read, substitutions that are following are now correctly extracted
- Reads (not substitutions) have been shifted by one. It now lines up with the sequence track and the bam file. See below. Please check.

![Screen Shot 2020-11-06 at 1 21 14 PM](https://user-images.githubusercontent.com/53857412/98401147-33d5c280-2033-11eb-9e58-aa64277c9930.png)
![Screen Shot 2020-11-02 at 12 01 04 PM](https://user-images.githubusercontent.com/53857412/98401153-36381c80-2033-11eb-8142-28d66675c4ce.png)

Read count information:
![Screen Shot 2020-11-06 at 1 25 15 PM](https://user-images.githubusercontent.com/53857412/98401396-9c24a400-2033-11eb-9655-7322d6f4c154.png)


- [ ] Update the version number in unpkg link in README.md if building a new release

pkerpedjiev commented on package.json:
```diff
@@ -1,6 +1,6 @@
 {
   "name": "higlass-pileup",
-  "version": "0.4.0",
+  "version": "0.4.1",
```
This should definitely be at least a minor version bump if not a major. We could just go ahead and release v1.0.0 with these changes.

alexander-veit replied:
Done

pkerpedjiev commented on src/PileupTrack.js:
```diff
@@ -188,9 +187,14 @@ const PileupTrack = (HGC, ...args) => {
   class PileupTrackClass extends HGC.tracks.Tiled1DPixiTrack {
     constructor(context, options) {
+      let baseUrl = `${getThisScriptLocation()}/`;
+      if (options.workerScriptLocation) {
```
Could you add this option to the README?

alexander-veit replied:
Done

pkerpedjiev commented on src/PileupTrack.js:
```diff
@@ -343,27 +361,27 @@ varying vec4 vColor;
 
       const fetchedTileKeys = Object.keys(this.fetchedTiles);
-      fetchedTileKeys.forEach((x) => {
+      fetchedTileKeys.forEach(x => {
```
Are you using an older version of prettier? Parenthesis around anonymous function arguments are on by default as of prettier v2.0

alexander-veit replied:
Thanks for noticing. I am using the correct version now.

pkerpedjiev commented on src/PileupTrack.js:
```diff
@@ -492,22 +511,29 @@ varying vec4 vColor;
       //   .range([0, this.dimensions[1]]);
       // HGC.utils.trackUtils.drawAxis(this, valueScale);
-      this.trackNotFoundText.text = 'Pete rules11!';
+      this.trackNotFoundText.text = 'Pete rules!!!';
```
Ummm... we should probably change this 🤣

alexander-veit replied:
I reluctantly changed it ;-)

pkerpedjiev commented on src/PileupTrack.js:
```diff
@@ -562,26 +588,54 @@ varying vec4 vColor;
+
         // var val = self.yScale.domain()[index];
+        if (
+          this.options.showReadCounts &&
```
I think what you refer to read counts is usually referred to as "coverage".

alexander-veit replied:
I changed it to coverage everywhere

pkerpedjiev commented on src/bam-fetcher-worker.js:
```diff
@@ -848,31 +952,50 @@ const renderSegments = (
-        addRect(xLeft, yTop, xRight - xLeft, height, PILEUP_COLOR_IXS.BG);
-
-        const substitutions = getSubstitutions(segment);
+        //console.log(xLeft, yTop, xRight - xLeft, height)
```
```suggestion
```

pkerpedjiev commented on src/bam-fetcher-worker.js:
```diff
@@ -848,31 +952,50 @@ const renderSegments = (
+        else{
+          addRect(xLeft, yTop, xRight - xLeft, height, PILEUP_COLOR_IXS.BG);
+        }
+        //console.log(allColors)
```
```suggestion
```

pkerpedjiev commented on src/bam-utils.js:
```diff
@@ -20,46 +24,74 @@ Object.keys(PILEUP_COLORS).map((x, i) => {
     }
   }
-
+  // if(!useCounts){
```
Might as well delete this, no?

pkerpedjiev commented on src/bam-utils.js:
```diff
@@ -20,46 +24,74 @@ Object.keys(PILEUP_COLORS).map((x, i) => {
 };
 
-export const getSubstitutions = (segment) => {
+export const getSubstitutions = (segment, seq) => {
```
What is seq? I've set a bad precedent here but docstrings would be good to have.

alexander-veit replied:
Done

pkerpedjiev commented on src/bam-utils.js:
```diff
@@ -82,29 +113,28 @@ export const getSubstitutions = (segment) => {
           pos: currPos,
           length: sub.length,
-          sub,
+          //sub,
```
```suggestion
```

pkerpedjiev commented on src/bam-utils.js:
```diff
@@ -82,29 +113,28 @@ export const getSubstitutions = (segment) => {
         currPos += sub.length;
       } else {
         // console.log('skipping:', sub.type);
+        // console.log(cigarSubs)
```
```suggestion
```

I know, I'm guilty as well but we should probably strive to get rid of useless console log statements.

alexander-veit replied:
Agreed.

pkerpedjiev reviewed (approved):
This is awesome! Thanks a ton for submitting this.

I added a bunch of minor comments but overall it looks great! Do you think there's any way to benchmark how fast it renders for a large-ish file and compare before these changes and after?

pkerpedjiev reviewed (commented):
Tried to change to comment rather than approve but apparently it doesn't work. I just wanted the review status to be set to `comment` to ensure that the little things get taken care of before merging.

alexander-veit commented:
Hey Pete, thanks a lot for the detailed review and the comments. I am out until Thursday, but will get started on the changes and the benchmark then right away.

alexander-veit commented:
I modified it according to your comments.

Furthermore, I added a new `segmentsToRow` function. It is a bit easier to read in my opinion, and resolves the 'overlapping reads' issue. Performance-wise it seems a bit faster than the old function, when you initially load the display (i.e. `prevSegments` is empty). I tested it with a window size of 150kb and 50,000 segments. When you zoom around it is  equal or a little bit slower than the old function (because of the way `prevSegments` is processed. However, in practice I could not notice any difference. If you experience this differently, please let me know.

Sorry, that I included it in this PR. It uses some of the modifications already made and it was easier that way. It should be easy to review, it is just an altered `segmentsToRow` function.

The coverage 'track' is currently a bit slow (it is switched off by default). That's not an issue when the window size is reasonable, but when you look at a 150kb window it takes around 80% of the `renderSegments` time. This is of course too much. If it's ok for you, I would like to address this in the next PR.

alexander-veit commented:
I have now a much faster version of the coverage track. I will create a PR for that once this PR is merged.

pkerpedjiev commented on src/index.html:
```diff
@@ -66,30 +67,127 @@
+            "type": "horizontal-sequence",
+            "width": 768,
+            "height": 25,
+            "server": "https://cgap-higlass.com/api/v1",
```
Not related to this PR, but why does the sequence track require a server if urls are provided in the `data` section?

alexander-veit replied:
I doesn't need it. I removed it.

pkerpedjiev commented on src/index.html:
```diff
@@ -98,8 +196,38 @@
+            "uid": "FylkvVBTSumoJ959HT4-5B",
+            "data": {
+              "type": "bam",
+              // Example Pete
```
You can take this out.

pkerpedjiev commented on src/index.html:
```diff
@@ -66,30 +67,127 @@
-        1069795.099783097,
-        1070508.6726894341
-      ],
+      // Example Pete
```
You can take this out :-)

pkerpedjiev reviewed (approved):
Looks great! Thanks a ton for all these updates 🙏
